RavenDB-21130 Corax null & empty string handling

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21130

### Additional description


Handling null values:
Each field has its own posting list dedicated solely to storing null values. This implementation seamlessly integrates into the existing indexing flow as follows:

Indexing proceeds as usual, with the NULL value represented as \0 during indexing (given that CompactTree doesn't support it).
For such a term, InsertTextualValues will:
--Create a new posting list (without support for small/single values) when the posting list doesn't exist.
--Utilize the standard update flow otherwise.

As we no longer store values, we must retain information within the EntryToTerm structure to indicate the presence of null values. Since 0b111 is already reserved to handle standard values (such as spatial, stored, or numeric), I opted to utilize the sign bit for this purpose. Normally, we store the term container ID within it (which serves as a page identifier in CompactTree). This allows us to determine the parent of the term (field compact tree). Since we're not storing it as a term we do not have such an identifier. In NULL case, I directly store the RootPage to identify which field requires an update.

### Type of change


- Optimization
- New feature

### How risky is the change?


- High

### Backward compatibility


- Breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.


### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
(using current test-base)

### Testing by RavenDB QA team

- This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.


### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

ayende committed:
RavenDB-21021 Include Corax DebugView (expose query structure) in Timings()

ayende committed:
RavenDB-21092 Corax streaming optimization - query with just order by

ayende committed:
RavenDB-20553 Now allow to define a compound field for indexes (not doing anything yet, just round tripping the information)

ayende committed:
RavenDB-20553 Generating automatic field from compound definition (early steps)

ayende committed:
RavenDB-20553 Small refactor to extract uncommon code outside of hot loop

ayende committed:
RavenDB-20553
- Adding support for most values in compound indexes
- For now, only supporting compound fields with 2 items

ayende committed:
RavenDB-20553
- Now compound indexes will use proper analyzer for the terms (but only a single term is supported)
- Can properly detect and optimize queries with simple equality and order by using a compound field

maciejaszyk committed:
RavenDB-21130 Introducing separate storage for posting list

ayende committed:
RavenDB-21137 Optimize compound sorting by fetching the first N+1 terms, then sorting just those, instead of the entire collection
- Allowing to limit the number of terms scanned during range operation

ayende committed:
RavenDB-21137 Fixing handling over very large terms in Corax when running equality comparisons

ayende committed:
RavenDB-21137  Fixing Studio compilation

maciejaszyk committed:
RavenDB-21130 Return NULL_VALUE in Exists primitive.

maciejaszyk committed:
RavenDB-21130 Introduce IncludeNull primitive to allow exposing null values in streaming query optimization

ayende committed:
RavenDB-21137 Avoid allocating a string and use a char buffer instead

ayende committed:
RavenDB-21137 Avoid allocating a string reader per text

ayende commented on src/Corax/Constants.cs:
```diff
@@ -9,11 +10,11 @@ public static class Constants
         
         
-        public const string EmptyString = "EMPTY_STRING";
+        public const string EmptyString = @"""";
```
I don't like this.
What happens when the user _wants_ to index the string `""` ?

ayende replied:
Shouldn't this be ␃ ?

ayende commented on src/Corax/IndexSearcher/IndexSearcher.TermMatch.cs:
```diff
@@ -219,12 +222,20 @@ public long NumberOfDocumentsUnderSpecificTerm<TData>(FieldMetadata binding, TDa
         
+        if (term is null)
+        {
+            //todo perf
```
I don't think this is something that we really should care, but I have been surprised before

ayende replied:
do we have any perf issues here?

ayende committed:
RavenDB-21137 Actually use a span instead of char to avoid nastiness in allocations

ayende committed:
RavenDB-21131 When writing compound fields values, postfix with offsets

arekpalinski commented:
Please check this failure:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/7841/testReport/junit/SlowTests.Issues/RavenDB_20529/CanCountNullAsTermInFacet_options___DatabaseMode___Single___SearchEngineMode___Corax_/

arekpalinski commented:
I scheduled another build and it crashed as well: https://jenkins.hibernatingrhinos.com/job/PR_Tests/7852/consoleFull

maciejaszyk committed:
RavenDB-21130 Decode postinglist in IncludeNullMatch

maciejaszyk committed:
RavenDB-21130 Debug method for EntriesModifications and EntryTermsWriter

maciejaszyk committed:
RavenDB-21021 Count real char size instead of whole structure to avoid writing "(CHAR)\0" in buffer.

maciejaszyk committed:
RavenDB-21190 Facet: use constant in case of NULL key | Null indexing: in case of frequency > 1 we've to add NullMarker again because shift deletes it in EntryToTerms.

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -1851,6 +1890,38 @@ void ThrowTriedToDeleteTermThatDoesNotExists()
+            long TryGetNullTermPostingList(CompactTree tree, out long termContainerId, out CompactTree.CompactKeyLookup keyLookup, out bool termExists, ref EntriesModifications entries, out bool alreadyIndexed)
+            {
+                keyLookup = default;
+                termContainerId = ~long.MaxValue | (tree.RootPage << 3); // sign bit means its null, rootpage identificates field
```
That may be a problem, since we are using varints here, no?
Which means that this would take a _lot_ of values.
Why do we need this to be anything but `0` ?

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -1851,6 +1890,38 @@ void ThrowTriedToDeleteTermThatDoesNotExists()
+
+                    ref var postingListState = ref MemoryMarshal.AsRef<PostingListState>(setSpace);
+                    {            entries.GetEncodedAdditions(_entriesAllocator, out var additions);
+
```
Formatting

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -1851,6 +1890,38 @@ void ThrowTriedToDeleteTermThatDoesNotExists()
+                alreadyIndexed = false;
+                //We want to obtain posting lists for NULL value from different tree than CompactTree to make rest code as valid ;-)
+                //We will
+                var postingList = _nullEntriesPostingLists.ReadInt64(indexedField.Name);
```
For null, we should _always_ just use a posting list.
Worst case scenario here is that we have a `null` posting list per field?
Not worth the effort to allocate for it.
Also means that we can _know_ what the value is statically.

ayende commented on src/Corax/Utils/EntryTermsReader.cs:
```diff
@@ -66,6 +66,13 @@ namespace Corax.Utils;
+///
+/// ### NULL Handling:
+/// In the context of NULL handling, we faced the challenge that the three lowest bits in the TermContainer were already in use.
+/// In order to store information about NULL terms, we utilized the highest bit (the sign bit) since container IDs are always positive.
```
That means that we are wasting a lot of space, however.

ayende replied:
Need to update this comment?

ayende committed:
RavenDB-21137 Moving from tree based to lookup persistent structure for managing container page lists, also managing the modified state _in memory_ and flushing on transaction commit only

ayende committed:
RavenDB-21137
- In memory "last container by page level metadata" to avoid searching too often
- Vectorized GetRequiredSizeAndPosition

ayende committed:
RavenDB-21137
- Remember the field root page instead of doing a dictionary lookup on each term

ayende committed:
RavenDB-21137 Avoid using a dictionary lookup (called millions of time) during recording of entries, instead, remember the index and use that to get the relevant entry

maciejaszyk committed:
RavenDB-21130 NULL handling in MoreLikeThis

ayende committed:
RavenDB-21137 Caching Term & Offset attributes from the analyzer to avoid dictionary lookups

ayende committed:
RavenDB-21131 Adding tests & fixing comments for compound fields

ayende committed:
RavenDB-21131 Allowing to optimize compound filter & sort

ayende committed:
RavenDB-21131 Fixing compound field queries that used multiple sort fields

maciejaszyk committed:
RavenDB-21130 Store RootPage as TermContainer without sign bit set to 1 to save storage space (varint compression)

maciejaszyk committed:
RavenDB-21130 Discard frequency from NULL posting list when performing batch fill on Exists query.

maciejaszyk committed:
RavenDB-21130 Do not insert NULL TermId into `EntriesForTerm`

ayende commented on src/Corax/Constants.cs:
```diff
@@ -1,28 +1,34 @@
         public const string DocumentBoost = "@document_boost";
-        
-        public static readonly Slice NullValueSlice, EmptyStringSlice, IndexMetadataSlice, DocumentBoostSlice, IndexTimeFieldsSlice;
+        public const string NullValueAsString = "NULL_VALUE";
```
Is this still used?

maciejaszyk replied:
We need it for `Facets` since null cannot be a key. We could use `\0` but `NULL_VALUE` in fit behaviour from Lucene and gives compatibility if somebody already expects NULLs under this term.

ayende commented on src/Corax/IndexSearcher/IndexSearcher.cs:
```diff
@@ -17,13 +15,14 @@
 using Corax.Queries.TermProviders;
 using Corax.Utils;
-using Sparrow.Compression;
+using NetTopologySuite.Triangulate.QuadEdge;
```
??

ayende commented on src/Corax/Utils/EntryTermsWriter.cs:
```diff
@@ -10,6 +11,8 @@ namespace Corax.Utils;
 
 public unsafe struct EntryTermsWriter : IDisposable
 {
+    public const long NullMarker = ~long.MaxValue;
```
still used?

ayende commented on src/Corax/Constants.cs:
```diff
@@ -82,7 +82,7 @@ static IndexWriter()
                     Slice.From(ctx, "MultipleTermsInField", ByteStringType.Immutable, out MultipleTermsInField);
                     Slice.From(ctx, "NullPostingLists", ByteStringType.Immutable, out NullPostingLists);
-
+                    Slice.From(ctx, "RootPages", ByteStringType.Immutable, out RootPagesSlice);
```
What is this used for?

maciejaszyk replied:
Leftover

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -1898,6 +1909,8 @@ private void InsertTextualField(Tree entriesToTermsTree, IndexedField indexedFie
 
                         dumper.WriteAddition(term, termId);
                         termContainerId = fieldTree.AddAfterTryGetNext(ref keyLookup, termId);
+                        if (termContainerId % Voron.Global.Constants.Storage.PageSize == 0)
+                            Debugger.Break();
```
should remove, no?

maciejaszyk replied:
Yes

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -1934,7 +1947,8 @@ private void InsertTextualField(Tree entriesToTermsTree, IndexedField indexedFie
                                 throw new ArgumentOutOfRangeException(entriesToTermResult.ToString());
                         }
-
+                        if (termContainerId % Voron.Global.Constants.Storage.PageSize == 0)
+                            Debugger.Break();
```
should remove?

ayende commented on src/Corax/Utils/EntryTermsReader.cs:
```diff
@@ -219,9 +230,10 @@ private void HandleRegularTerm(long termContainerId)
         }
 
-        if ((termContainerId & EntryTermsWriter.NullMarker) != 0 ) // null value
+        if ((TermId % Voron.Global.Constants.Storage.PageSize) == 0 && _rootPages.ToSpan().BinarySearch(TermId / Voron.Global.Constants.Storage.PageSize) >= 0) // null value
```
Wouldn't this be better as a `HashSet`?

maciejaszyk replied:
Why? We also use the reader for `UnaryMatches`/`facets`/`morelikethis`, so this list can be created a lot.` NativeList` gives us little GC pressure in comparison to `HashSet,` right?

maciejaszyk committed:
RavenDB-21130
- Better naming of constants
- Empty string is EndOfText character
- Exists will not decode values in Fill since we're performing it in MultiTermMatch
- Removing debug leftovers from IndexWriter
- Removing of unused usings in IndexSearcher

arekpalinski commented:
test this in debug please

maciejaszyk committed:
RavenDB-21137
- Copy compund fields in index definition
- In case of writing stored NULL we've to read RootPage first
- CoraxQueryBuilder: when we decide to stream manually we've to skip SortingMatch wrapper.
